mm2src/coins_activation/src/solana_with_tokens_activation.rs

```diff
@@ -0,0 +1,228 @@
+use crate::platform_coin_with_tokens::{EnablePlatformCoinWithTokensError, GetPlatformBalance,
+                                       InitTokensAsMmCoinsError, PlatformWithTokensActivationOps, RegisterTokenInfo,
+                                       TokenActivationParams, TokenActivationRequest, TokenAsMmCoinInitializer,
+                                       TokenInitializer, TokenOf};
+use crate::prelude::*;
+use crate::prelude::{CoinAddressInfo, TokenBalances, TryFromCoinProtocol, TxHistoryEnabled};
+use crate::spl_token_activation::SplActivationRequest;
+use async_trait::async_trait;
+use coins::my_tx_history_v2::TxHistoryStorage;
+use coins::solana::spl::{SplProtocolConf, SplTokenCreationError};
+use coins::{solana_coin_from_conf_and_params, BalanceError, CoinBalance, CoinProtocol, MarketCoinOps,
+            SolanaActivationParams, SolanaCoin, SplToken};
+use common::mm_ctx::MmArc;
+use common::mm_error::prelude::MapToMmResult;
+use common::mm_error::MmError;
+use common::mm_metrics::MetricsArc;
+use common::mm_number::BigDecimal;
+use common::Future01CompatExt;
+use futures::future::AbortHandle;
+use serde_derive::{Deserialize, Serialize};
+use serde_json::Value as Json;
+use std::collections::HashMap;
+
+pub struct SplTokenInitializer {
+    platform_coin: SolanaCoin,
+}
+
+impl TokenOf for SplToken {
+    type PlatformCoin = SolanaCoin;
+}
+
+#[async_trait]
+impl TokenInitializer for SplTokenInitializer {
+    type Token = SplToken;
+    type TokenActivationRequest = SplActivationRequest;
+    type TokenProtocol = SplProtocolConf;
+    type InitTokensError = SplTokenCreationError;
+
+    fn tokens_requests_from_platform_request(
+        platform_params: &SolanaWithTokensActivationRequest,
+    ) -> Vec<TokenActivationRequest<Self::TokenActivationRequest>> {
+        platform_params.spl_tokens_requests.clone()
+    }
+
+    async fn enable_tokens(
+        &self,
+        activation_params: Vec<TokenActivationParams<SplActivationRequest, SplProtocolConf>>,
+    ) -> Result<Vec<SplToken>, MmError<Self::InitTokensError>> {
+        let mut tokens = Vec::new();
+        for param in activation_params.into_iter() {
+            tokens.push(SplToken::new(
+                param.protocol.decimals,
+                param.ticker,
+                param.protocol.token_contract_address,
+                self.platform_coin.clone(),
+            )?)
```

Comment:
was not able to use `map()` - if there is a way to avoid normal loop iteration for iterator + algorithm please let me know

Reply:
You can collect items into `Result<Vec<_>, _>` and use the `?` operator.
```rust
let tokens = activation_params.into_iter().map(|param| SplToken::new(
                param.protocol.decimals,
                param.ticker,
                param.protocol.token_contract_address,
                self.platform_coin.clone(),
            )).collect::<Result<Vec<_>, _>>()?;
```